### PR TITLE
plat: intel: Add BL31 support to Intel Stratix10 SoCFPGA platform

### PR DESCRIPTION
**Pull request by thloh85-intel**
This adds BL31 support to Intel Stratix10 SoCFPGA platform. BL31 in TF-A
supports:
- PSCI calls to enable 4 CPU cores
- PSCI mailbox calls for FPGA reconfiguration

Signed-off-by: Loh Tien Hock <tien.hock.loh@intel.com>

### CODE REVIEW

File: plat/intel/soc/stratix10/bl31_plat_setup.c [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			  &bl33_image_ep_info : &bl32_image_ep_info;
+
+	/* None of the images on this platform can have 0x0 as the entrypoint */
+	if (next_image_info->pc)
+		return next_image_info;
+	else
+		return NULL;
+}
+
+#if LOAD_IMAGE_V2
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/bl31_plat_setup.c]: LOAD_IMAGE_V2 has been removed, so you can remove the part of the code you don't use from here.

File: plat/intel/soc/stratix10/include/s10_mailbox.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
+#ifndef __S10_MBOX__
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/include/s10_mailbox.h]: This file doesn't have a license header.

File: plat/intel/soc/stratix10/include/s10_mailbox.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+#define MBOX_OFFSET		0xffa30000
+
+#define MBOX_ATF_CLIENT_ID	0x1
+#define MBOX_JOB_ID		0x1
+
+/* Mailbox interrupt flags and masks */
+#define MBOX_INT_FLAG_COE	0x1
+#define MBOX_INT_FLAG_RIE	0x2
+#define MBOX_INT_FLAG_UAE	0x100
+#define MBOX_COE_BIT(INTERRUPT)	(INTERRUPT & 0x3)
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/include/s10_mailbox.h]: Guard macro arguments inside (). There are a few occurrences here between lines 13 and 29.

File: plat/intel/soc/stratix10/plat_psci.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+ */
+
+#include <arch_helpers.h>
+#include <assert.h>
+#include <common/debug.h>
+#include <errno.h>
+#include <mmio.h>
+#include <drivers/arm/gic_common.h>
+#include <drivers/arm/gicv2.h>
+//#include <plat_arm.h>
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_psci.c]: Remove commented-out line.

File: plat/intel/soc/stratix10/plat_psci.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+void plat_pwr_domain_on_finish(const psci_power_state_t *target_state)
+{
+	for (size_t i = 0; i <= PLAT_MAX_PWR_LVL; i++)
+		VERBOSE("%s: target_state->pwr_domain_state[%lu]=%x\n",
+			__func__, i, target_state->pwr_domain_state[i]);
+
+	/* Program the gic per-cpu distributor or re-distributor interface */
+	gicv2_pcpu_distif_init();
+	gicv2_set_pe_target_mask(plat_my_core_pos());
+//	plat_arm_gic_pcpu_init();
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_psci.c]: Remove commented-out code. Same in the functions below.

File: plat/intel/soc/stratix10/plat_sip_svc.c [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+			}
+		}
+	}
+
+	return -1;
+}
+
+unsigned int address_in_ddr(uint32_t *addr)
+{
+	if (((long long)addr > DRAM_BASE) &&
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_sip_svc.c]: Use ``uintptr_t`` or ``unsigned long long`` for addresses, not signed integers.

File: plat/intel/soc/stratix10/plat_psci.c [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+/*******************************************************************************
+ * plat handlers to shutdown/reboot the system
+ ******************************************************************************/
+static void __dead2 plat_system_off(void)
+{
+	wfi();
+	ERROR("System Off: operation not handled.\n");
+	/* disable coherency */
+	/*plat_arm_interconnect_exit_coherency();*/
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_psci.c]: nit: remove or replace by a TODO

File: plat/intel/soc/stratix10/plat_sip_svc.c [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+/* Total buffer the driver can hold */
+#define FPGA_CONFIG_BUFFER_SIZE 4
+
+int current_block;
+int current_buffer;
+int current_id = 1;
+int max_blocks;
+uint32_t bytes_per_block;
+uint32_t blocks_submitted;
+uint32_t blocks_completed;
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_sip_svc.c]: Can you make them static?

File: plat/intel/soc/stratix10/plat_sip_svc.c [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+				   uint64_t x4,
+				   void *cookie,
+				   void *handle,
+				   uint64_t flags)
+{
+	ERROR("%s: unhandled SMC (0x%x)\n", __func__, smc_fid);
+	SMC_RET1(handle, SMC_UNK);
+}
+
+struct fpga_config_info fpga_config_buffers[FPGA_CONFIG_BUFFER_SIZE];
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/plat_sip_svc.c]: Static?

**Comment by thloh85-intel** [comment]
> There are some checkpatch messages:
> 
> ```
> plat/intel/soc/stratix10/soc/s10_mailbox.c:11: ERROR: open brace '{' following function definitions go on the next line
> +static int fill_mailbox_circular_buffer(uint32_t header_cmd, uint32_t *args,
> +                                       int len) {
> 
> plat/intel/soc/stratix10/soc/s10_mailbox.c:53: WARNING: suspect code indent for conditional statements (16, 16)
> +               if (timeout-- < 0)
> +               return MBOX_NO_RESPONSE;
> ```
> Also, could you try to build with `ERROR_DEPRECATED=1` and do the changes needed to fix the build?

For some odd reason, my run of checkpatch didn't catch the error :( 
I'll fix this, and will run with ERROR_DEPRECATED=1. Thanks!

File: plat/intel/soc/stratix10/plat_sip_svc.c (hunk 8 above)
Comment (thloh85-intel) [reply plat/intel/soc/stratix10/plat_sip_svc.c]: Hmm why make them static?
This should be used only locally in this c file, not outside of it.

**Comment by thloh85-intel** [comment]
> > There are some checkpatch messages:
> > ```
> > plat/intel/soc/stratix10/soc/s10_mailbox.c:11: ERROR: open brace '{' following function definitions go on the next line
> > +static int fill_mailbox_circular_buffer(uint32_t header_cmd, uint32_t *args,
> > +                                       int len) {
> > 
> > plat/intel/soc/stratix10/soc/s10_mailbox.c:53: WARNING: suspect code indent for conditional statements (16, 16)
> > +               if (timeout-- < 0)
> > +               return MBOX_NO_RESPONSE;
> > ```
> > Also, could you try to build with `ERROR_DEPRECATED=1` and do the changes needed to fix the build?
> 
> For some odd reason, my run of checkpatch didn't catch the error :(
> I'll fix this, and will run with ERROR_DEPRECATED=1. Thanks!

OK it appears I used old copy of checkpatch.pl which had bugs that caused the error to be escaped. Now using the latest checkpatch.pl and I caught those errors. Thanks!

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
Don't worry about the CI failure, it's because of the FVP port merged last week and are not present in your branch.

**Pull request merged**